Add a Semantic Convention for Serverless workload (#8)
dynatrace-oss-contrib/opentelemetry-specification: specification/data-resource-semantic-conventions.md

| container.image.name | Name of the image the container was built on. | `gcr.io/opentelemetry/operator` |
| container.image.tag | Container image tag. | `0.1` |

### Function as a Service

**type:** `faas`

**Description:** An instance of a function that runs on a platform where underlying hardware concerns, such as provisioning and scaling, have been abstracted away by the platform.

| Label  | Description  | Example  | Required? |
|---|---|---|--|
| faas.name | The name of the function being executed. | `my-function` | Yes |

> | faas.id | The unique name of the function being executed. <br /> For example, in AWS Lambda this field corresponds to the [ARN] value, in GCP to the URI of the resource, and in Azure to the [FunctionDirectory] field. | `arn:aws:lambda:us-west-2:123456789012:function:my-function` | Yes |

Reviewer: ```suggestion
| faas.id | The unique id of the function being executed. <br /> For example, in AWS Lambda this field corresponds to the [ARN] value, in GCP to the URI of the resource, and in Azure to the [FunctionDirectory] field. | `arn:aws:lambda:us-west-2:123456789012:function:my-function` | Yes |
```

> | faas.identifier | The unique name of the function being executed. <br /> For example, in AWS Lambda this field corresponds to the [ARN] value, in GCP to the URI of the resource, and in Azure to the [FunctionDirectory] field. | `arn:aws:lambda:us-west-2:123456789012:function:my-function` | Yes |

Reviewer 2: Propose shorten this to `faas.id`. See above for my suggestion to rename the old `faas.id` to `faas.execution`.

> | faas.service | Platform name where the function is executed, e.g. `aws-lambda`, `cloud-functions`, `azure-functions`, ... | `aws-lambda` | Yes |

Reviewer 2: Propose some way to make sure these are as consistent as possible. Possibly a list of known values so far? It should be flexible enough to allow a new service we don't currently know about to be used here, but it should not be so flexible that different SIGs use different names for the same service. For instance, one SIG may put `cloud-functions` and another might put `google-cloud-functions`.

Author: Looking at the resource list [(link)](https://github.com/dynatrace-oss-contrib/opentelemetry-specification/blob/master/specification/data-resource-semantic-conventions.md#cloud), there is the *cloud* convention which partially cover the purpose of this attribute. However, I am thinking that this level of granularity is not needed given the problems you mentioned.

Reviewer 3: Are you planning to change this then? E.g. move it to "cloud.service" or something like that?

Author: I am more keen on completely remove it. The original idea was to allow services that are build on top of FaaS (e.g. serverless, claudia.js, or even the multiple different implementations offered by azure) to tag the resource with their own identifier. I am not sure anymore that this fine grained level is necessary.

Reviewer 4: Agreed, let's remove that.

| faas.version | The version string of the function being executed as defined in [Version Attributes]. | `semver:2.0.0` | No |

> | faas.instance | String containing the execution environment ID. | `my-function:instance-0001` | No |

Reviewer 3: ```suggestion
| faas.instance | The execution environment ID as a string. | `my-function:instance-0001` | No |
```

| container.image.name | Name of the image the container was built on. | `gcr.io/opentelemetry/operator` |
| container.image.tag | Container image tag. | `0.1` |

### Function as a Service

**type:** `faas`

**Description:** An instance of a function that runs on a platform where underlying hardware concerns, such as provisioning and scaling, have been abstracted away by the platform.

| Label  | Description  | Example  | Required? |
|---|---|---|--|
| faas.name | The name of the function being executed. | `my-function` | Yes |

> <<<<<<< HEAD

Reviewer 2: you checked in git resolution markers

Author: Damn!! IntelliJ made me lose the [ctrl|cmd]+s muscle memory :smile:

| container.image.name | Name of the image the container was built on. | `gcr.io/opentelemetry/operator` |
| container.image.tag | Container image tag. | `0.1` |

### Function as a Service

**type:** `faas`

> **Description:** A serverless instance.

Reviewer 2: Because serverless is more of a buzzwordy term and not very well defined, I would try to think of some alternative phrasing.
```suggestion
**Description:** An instance of a function that runs on a platform where underlying hardware concerns such as provisioning and scaling have been abstracted away by the platform.
```

Reviewer 4: You could still add (aka "serverless").

Author: I agree, otherwise totally different products, such as knative, would be put into this category.

| Label  | Description  | Example  | Required |
|---|---|---|--|

> | faas.region| A specific geographical location where the function is executed. | `us-east-2` | Yes |

Reviewer: As this is mandatory it's needed to add some hint what to add in case a function doesn't know where it is running. Usually there is no need for a function to know it's region, it's not even needed that a FaaS provider offers a concept region.

> | faas.name | The name of the function being executed. | `opentelemetry-lambda` | Yes |

Reviewer 2: Using "opentelemetry" in the example name of an end user's application may confuse people and make them think you're referring to the instrumentation library name. I would change this to be something totally unrelated like `widget-provider`.

> | faas.identifier | The unique name of the function being executed. <br /> For example, in AWS this field correspond to the [ARN] value, in GCP to the URI of the resource, and Azure to the [FunctionDirectory] field. | `opentelemetry-lambda` | Yes |

Reviewer 2: ```suggestion
| faas.identifier | The unique name of the function being executed. <br /> For example, in AWS Lambda this field corresponds to the [ARN] value, in GCP to the URI of the resource, and in Azure to the [FunctionDirectory] field. | `opentelemetry-lambda` | Yes |
```

Reviewer 2: Same comment on naming as from `faas.name`

> | faas.identifier | The unique name of the function being executed. <br /> For example, in AWS this field correspond to the [ARN] value, in GCP to the URI of the resource, and Azure to the [FunctionDirectory] field. | `opentelemetry-lambda` | Yes |

Reviewer: `provider` is optional but `identifier` is mandatory. We may end up in a quite useless string here as the format is unknown. if we know it's AWS lambda and this is an ARN then it's possible to parse it.

> | faas.version | The version of the function being executed. | `opentelemetry-latest` | No |

Reviewer 2: Versions in the spec typically follow the format `semver:0.1.2`, `git:a4c2ef`, etc.

Reviewer 2: The description leaves it ambiguous to what is meant by version. Is this the version of the lambda service, the version of the end user's application, or the version of the telemetry library?

Author: I agree, we should emphasize the semantic versioning. Furthermore, this comment made me thinking... Should we also cover information about the runtime in which the function is executed? For instance OS, language runtime version, etc..

Reviewer 4: I think the latter (OS, ...) are part of resources and the semantic conventions around those.

> | faas.provider | Provider name of the function, e.g. `aws`, `gpc`, `azure`, `serverless`, ... | `aws` | No |

Reviewer 2: What do you do if AWS has a second FaaS service which competes with Lambda? Azure already has multiple different types of function service. Is this meant to be the name of the Platform, or of the service used?

Reviewer 4: I don't think this should be optional.

Author: My idea was to cover the platform, but from an end-user perspective maybe the name of the function service is better. This way, end-users can aggregate trace data by function service and not just by platform.  WDYT?

Reviewer 3: This smells like it could be more general like, `cloud.provider=aws`, `cloud.service=lambda`.

Reviewer 5: `gpc` -> `gcp`?

> | faas.payload | The input passed to the function. | `{name: 'opentelemetry'}` | No |

Reviewer 2: This description leaves a lot of room for interpretation. For instance, on AWS Lambda triggered by HTTP, is this the whole `event` object or just the POST body?

Reviewer 4: This could get very big potentially.
Also, we have to consider masking data here (i.e. make sure the server does it)

Author: Agree, I should make it more specific.

Author: > This could get very big potentially.
> Also, we have to consider masking data here (i.e. make sure the server does it)

On a second thought, is this field really important? Having this field might be dangerous, from privacy and resource required POV

Reviewer 3: True. There was talk about having "enhanced information"/"discarded by default" fields here: https://github.com/open-telemetry/opentelemetry-specification/pull/397#discussion_r367966835 But for now I think it would be best to leave the field out.

> | faas.trigger | Type of the trigger that spawned the function, e.g. `http`, `S3`, `PubSub`, ... | `http` | Yes |

Reviewer 2: `S3` is a service, but `http` is a transport. If you want this to be the transport used, then I would remove the `S3` trademark name and use all generic transports like http and PubSub. If you want this to be the service that triggered the function, then I would use examples that are all service names like API Gateway, S3, SQS.

Author: Makes sense! Maybe an enum to chose from is better. Say: http, pubsub, event, timer, manual

Reviewer: The function may not know what was the actual trigger. This could be completely abstracted by the FaaS platform (e.g. harmonize http/https and http2).

Reviewer 4: Let's make this optional, then.

Author: I would keep it mandatory. It is true that the function might not know the exact protocol used (e.g. http or https), but conceptually the function has the knowledge if it answers to an HTTP request, to a PubSub event, to a cron job, and so forth.

Reviewer 4: Fair enough, if we are confident that this information is available to the instrumentation code, let's keep it mandatory.

> | faas.trigger | Type of the trigger that spawned the function, e.g. `http`, `S3`, `PubSub`, ... | `http` | Yes |

Reviewer 3: Nitpick: I would sort all required attributes to the top.

> | faas.instance.id | String containing the unique execution id. | `opentelemetry-0001` | Yes |

Reviewer 2: Is this instance id or execution id? An instance can be reused for multiple executions. What do you do if the service provider does not make this information available to the function?

Reviewer 2: Same comment on naming.

Reviewer 4: This should be optional (see @dyladan 's comment).

Author: I didn't think about this. You right, this should be execution id.

Reviewer 2: You still may want instance ID in the case where you save some state to a global variable which is saved between invocations. If your first invocation works, but the second one fails consistently, this might be good information to have for debugging.

Author: That's a good point!

> 

Reviewer 3: That many newlines?

| container.image.name | Name of the image the container was built on. | `gcr.io/opentelemetry/operator` |
| container.image.tag | Container image tag. | `0.1` |

### Function as a Service

**type:** `faas`

**Description:** An instance of a function that runs on a platform where underlying hardware concerns, such as provisioning and scaling, have been abstracted away by the platform.

| Label  | Description  | Example  | Required? |
|---|---|---|--|
| faas.name | The name of the function being executed. | `my-function` | Yes |
| faas.id | The unique name of the function being executed. <br /> For example, in AWS Lambda this field corresponds to the [ARN] value, in GCP to the URI of the resource, and in Azure to the [FunctionDirectory] field. | `arn:aws:lambda:us-west-2:123456789012:function:my-function` | Yes |
| faas.version | The version string of the function being executed as defined in [Version Attributes]. | `semver:2.0.0` | No |

> | faas.instance | The execution environment ID as a string. | `my-function:instance-0001` | No |

Reviewer 6: What is the use case for this field? Typically, faas instance is not of interest (it might be a short living thingy). The description does not tell about the nature of the ID (uniqueness etc.).

| cloud.account.id | The cloud account id used to identify different entities. | `opentelemetry` |
| cloud.region | A specific geographical location where different entities can run | `us-central1` |
| cloud.zone | Zones are a sub set of the region connected through low-latency links.<br/> In aws it is called availability-zone. | `us-central1-a` |

> [ARN]:https://docs.aws.amazon.com/general/latest/gr/aws-arns-and-namespaces.html

Reviewer 3: ```suggestion
[ARN]: https://docs.aws.amazon.com/general/latest/gr/aws-arns-and-namespaces.html
```
